### DISCUSSION TITLE
New PCB

### DISCUSSION THREAD

**Discussion by Dakkaron** [Dakkaron/Fairberry#10] [Ideas]
This is a discussion around the new PCB which should replace the Arduino and the adapter PCB.

The point of this new PCB is to reduce the necessary manual soldering and also increase durability of the circuit.

In issue https://github.com/Dakkaron/Fairberry/issues/6 @dumblob and I suggested the following features of the new PCB:

- Capability to use an USB-C, Micro USB or reversed Micro USB plug
- Plug soldered directly to the PCB or connected via wires to through-holes on the PCB
- Should be as small as possible to not waste space

A few more things from my side:
- SPI interface should be exposed
- USB pass-through charging?

What other ideas do you have?

**Comment by dumblob** [comment]
I've given it a quick thought and for my use case I am kind of indifferent on whether to have a USB pass-through or not.

The pro of having a pass-through USB is I don't need to use wireless charging (which I'm not the biggest fan of, but it's not a deal breaker for me) and can charge my phones as I'm used to.

The con is I have to put off the case - but this shall be easy if we end up with "flexible parts" made from a sturdy non-flexible material. The pro is it won't affect power consumption and it'll be a little cheaper to buy parts for and a little less trouble to assemble (but we should anyway aim for automated assembly & soldering of ICs and small parts).

So I'd probably consider the power demand and if that was really negligible as a second parameter I'd consider the size of the PCB - if it would allow things mentioned below.

Regarding SPI I agree - though only as an additional interface and it shouldn't replace the usual over-USB flashing (which can potentially be done also from the phone itself - e.g. changing the layout - imagine using Termux on Android or [iSH](https://github.com/ish-app/ish/ ) on iPhones wouldn't even require one to mess with Android SDK).

Oh and if the PCB space will allow for it, then I'd just put there the most capable unused GPIOs as their break out in the form of hand-solderable pads (or even holes if there already will be any other holes on the PCB - I don't want to increase the complexity nor price). One never knows how and when this might come handy. This includes GPIOs from Arduino but maybe also unused ports from the BB keyboard itself.

Last but not least the PCB could be kind of "universal" and thus maybe allow for certain non-negligible leeway (depending on the PCB size, but my guess is 10 mm should be enough) with regards to the BB keyboard piece to accommodate for different USB port positions (in all 3 axes) as they're not always in the middle. The rule of thumb would be to "shrink" the PCB from full width:
![PCB-full_width](https://user-images.githubusercontent.com/2451044/167482687-026d6680-180c-4fbf-ad11-bac345ceefa8.jpg)
to shrinked width:
![PCB-shrinked_width](https://user-images.githubusercontent.com/2451044/167482761-7cb7c581-5a07-436d-b042-bd04c6c2da32.jpg)

IDK, this is all just quick ideas.

**Reply by Dakkaron** [reply]
Sorry for the late answer, I thought I had already answered. Too many places to check ;)

Since there were some developments on the case front, also the requirements for the new PCB could change.

In the newest design, I screwed the keyboard to the case. This makes removal/attachment really fast and easy, since I just have to slide off the bottom part. Because of this, I would drop pass-through charging from the list of requirements, since it's really complicated to implement and it's actually pretty easy to remove the keyboard for charging.

Adding pass-through charging would take quite a bit of effort, and there is always a bit of risk, since if it's implemented wrong, it might fry the keyboard's electronics. So I am fine now with removing that requirement.

> Regarding SPI I agree - though only as an additional interface and it shouldn't replace the usual over-USB flashing (which can potentially be done also from the phone itself - e.g. changing the layout - imagine using Termux on Android or [iSH](https://github.com/ish-app/ish/) on iPhones wouldn't even require one to mess with Android SDK).

Sure, the SPI interface is only for emergencies and/or initial flashing. I guess, if I get the board made from a place like JLCPCB, the ATMega32u4 will probably not have the Arduino Bootlader preinstalled, so there needs to be a way to get it on the chip in the first place.

I had a look at flashing from the phone. There are a few Arduino IDEs for Android. The first one plainly didn't work with the board (it requires the Reset button to be pressed during flashing, but the Arduino Pro Micro doesn't have one). There is another one that looks better, but I haven't really tested it. But that one is pretty complicated to setup, so I am not sure if that's something that I would generally recommend to beginners.

> Oh and if the PCB space will allow for it, then I'd just put there the most capable unused GPIOs as their break out in the form of hand-solderable pads (or even holes if there already will be any other holes on the PCB - I don't want to increase the complexity nor price). One never knows how and when this might come handy. This includes GPIOs from Arduino but maybe also unused ports from the BB keyboard itself.

Good idea! We already need holes for the USB wires, It shouldn't increase the price by too much. Assembly alone will probably already cost ~€15 per piece, so another euro or so for the PCB wouldn't be too much.

**Comment by Dakkaron** [comment]
![grafik](https://user-images.githubusercontent.com/6311558/168932028-7ef6da20-cae0-48e9-88c9-2e0dc6bb5ea4.png)

This is supposed to be the first version of the PCB schematic. Do you all have any feedback?

Go easy on me, it's the first time I've done something like this.

PS: One thing I just noticed: I probably could have separated out a decently usable SPI- and I2C header (meaning signal wires + VCC + GND). Right now it's all bunched up rather randomly...

**Reply by spARTacusFromCB** [reply]
That is kind of a physical representation of the current as-is design isn't it, the current arrangement with the separate Beakout and Arduino boards?

**Reply by Dakkaron** [reply]
This is the circuit diagram (yes, heavily inspired (not to call it stealing) from the Arduino and the Breakout that I currently use) for the upcoming All-in-one PCB.

I also found a service (pcbway) that would possibly be able to assemble the whole thing (minus the USB port, but that's really easy soldering).

I requested the Hirose connector to be added to JLCPCB's library, so they can assemble that, and they did, but only for chinese customers... Is anyone here from China?

**Reply by dumblob** [reply]
@Dakkaron thanks for the update and all the effort!

I am a bit puzzled JLCPCB added the connector only for Chinese customers - did they tell you why this weird artificial restriction?

**Reply by Dakkaron** [reply]
Yeah, I was puzzled as well^^ Apparently, they have two different production lines, one for domestic customers and one for international customers. And they only added the connector to their domestic production line. That support guy didn't have more information on that. He said, he'll try to ask them to also add it to the international line, but he's not sure if they will do it and how long it would take.

PCBWay on the other hand is much more free on the parts they allow customers to use, but I don't have any experience with them.
They seem to have some kind of offer that makes 20 PCBs quite cheap. So if we get 20 people together, who commit to actually buying the parts (and I manage to finish the PCB design), then we should be good to go.

**Reply by spARTacusFromCB** [reply]
I guess I can't easily enough "read" the circuit diagram (deficiency on my part) as I was kind of looking for the Horise socket on it. I guess I do now see that it doesn't represent the current two board prototype, and the indication on the diagram for the "Q10 KB" is where the Horise connector will reside?

I think there was probably a good handful of people in the CB threads that indicated either by post reply or like that they were interested. However, I fear some of those were referring to a finished product. I'd still be good with just hardware, as a first step after the PCBWay or JLCPCB efforts at the simplified PCB approach. However, some will have difficulty with envisioning the step to the finished product, without them seeing it. A challenge to perhaps somehow overcome.

If you are thinking that 20 people need to commit for a certain step, you probably have some math numbers already in mind. Do you think it's time to start something like a campaign or a kickstarter, or just more directly with/to yourself?

**Reply by Dakkaron** [reply]
The circuit diagram is just a logical representation, it doesn't show the actual locations.

Yes, BBQ10KBD is the Hirose socket. I imported the part from somebody else's project, they did the naming.

Regarding numbers: You are right, I did some numbers. I think, even though PCBWay would do the whole thing and JLCPCB wouldn't do the connector, I'd probably first do a trial batch on JLCPCB and hand-solder the connector, since I can do much lower numbers on JLCPCB. I can still do a bigger run later on PCBWay, but first I need to verify that the prototype works.

JLCPCBs prices have much less scaling in them than PCBWay. So smaller quantities are more viable there.

**Reply by Dakkaron** [reply]
![grafik](https://user-images.githubusercontent.com/6311558/169405560-38a0d97d-6483-4dda-9732-45760f568d9d.png)

I made it! This can now be ordered on JLCPCB. Sadly, without the Hirose connector soldered to it, that still needs to be done. Also the USB connector needs soldering. But I've done a few of these Hirose connectors so far, and by now I can solder them pretty reliably.

So that puts me at a point, where I could make those for around €75 per piece including my labour cost. This is starting to look very viable. I'll have to test the whole process and see how much work is really there to do, to judge, what kind of price I really have to ask for.

At least for now, though, I would rather avoid going to Kickstarter, since that adds a whole rat's nest of problems...

**Reply by spARTacusFromCB** [reply]
What will "it" be expected to look like and size in terms of how big the new PCB/board will be and how that will have to be arranged with also the Keyboard and USB, or into a bottom case?

**Reply by Dakkaron** [reply]
In the largest dimensions its 66mmx13mm. That is small enough to fit in the bottom case next to the USB plug mount (which takes a lot of space and is the main cause for the weird squeeze that I had to do around the center portion of the board).

So that way, it is possible to have the keyboard lie flat on the screen and still fit the whole board in there.

**Reply by spARTacusFromCB** [reply]
So the USB connector (if hard soldered) would be in the centre of the board, creating almost a T when plugged into the bottom of the phone? Then, the Q10 keyboard would be offset further up/on the phone to the max extent permitted by the ribbon cable length and bends? Or, if the keyboard was desired to not be on the phone but at the bottom of it and with the keyboard surface almost same plane as phone glass, then the keyboard would be stacked/layered above the T? Did I get any of that correct?

**Reply by spARTacusFromCB** [reply]
> What will "it" be expected to look like and size in terms of how big the new PCB/board will be and how that will have to be arranged with also the Keyboard and USB, or into a bottom case?

I now realize you included a picture (based on you also including that on CB). The picture didn't originally come through well on my phone and I missed right over it. Sorry.

**Reply by spARTacusFromCB** [reply]
I think I read "SPI" and "I2C" on the board. So it would support those?

**Reply by Dakkaron** [reply]
For now I didn't provision a position to hard-solder the USB port. I thought about it, especially due to the input from @dumblob. But I think it would be pretty hard to line-up the PCB with the phone's USB port. Let's see, maybe we'll come up with something satisfactory and add the port later.

I found a mistake in my circuit and will have to redo the layout... Currently, the clock generating circuit is too far away from the microprocessor.

>I think I read "SPI" and "I2C" on the board. So it would support those?

Yes. They aren't used by the keyboard, so I just routed the pins to the outside, so that they can be used for some kind of expansion, if so desired.

>Then, the Q10 keyboard would be offset further up/on the phone to the max extent permitted by the ribbon cable length and bends?

The ribbon cable is really long, so the keyboard can be in almost any position along the length of the phone.

> Or, if the keyboard was desired to not be on the phone but at the bottom of it and with the keyboard surface almost same plane as phone glass, then the keyboard would be stacked/layered above the T? Did I get any of that correct?

Yes, both options would be possible. Mostly, because the USB is connected by wire and thus there is a lot of flexibility in positioning on that side as well.

**Reply by dumblob** [reply]
Very nice!

Btw. it seems it could be made even narrower to fit e.g. small variants of Apple iPhones (which will have USB due to EU regulation and could become a good portion of your future "customer base").

So while redoing the layout, could you make it narrower so that it's max 56 mm long?

**Reply by Dakkaron** [reply]
That's a good point, haven't considered that.

One point to consider, though, is that the keyboard itself is around 63mm wide. So shrinking the board below that width won't make an impact... Owners of very narrow phones will, unfortunately, have to live with an oversized attachment...

**Reply by dumblob** [reply]
>One point to consider, though, is that the keyboard itself is around 63mm wide.

Yep. Still it's much better to have only one physical part be the bottle neck than multiple. Note also that the PCB itself shouldn't be wider than ~60 to leave clarance for the plastic case "walls". Ideally though it shall be even narrower to ensure some more clearance for "weird" (by us here totally unexpected) ways of bending the flex cable next to the hirose connector.

PCB layouters hate it, but as I'm a prick, I'll say it again - better make it even smaller than be lazy :wink:.

**Reply by Dakkaron** [reply]
It's honestly going to be really hard to make this smaller, without dropping ports. The biggest problem is the vertical space, which is very limited, so that the keyboard can be placed on top of the phone without the PCB in between. With the space needed for the USB connector clamp, there is just not a lot of space in the Y direction.

In the design I showed before, I actually had so little space left, that I had to move the clock generator to the left part, which won't work since the distance between it and the microcontroller is too large. So I'm redoing the layouting, and moving the generator to the right. For this, the right section has to be even bigger (I increased the size in the Y dimension for that).

Getting this smaller is only possible if I drop some of the ports.

I can't drop SPI, since I'll need it for the initial programming of the board (since I guess they won't come with the Arduino bootloader from the factory).

I don't want to drop I2C, since it's a really easy and simple way to extend the device's functionality if needed.

I don't want to drop the analog pins, since they would allow for stuff like a trackpoint.

I could drop the one "conenctor" where I put VCC, GND and a single digital pin. (VCC and GND are replicated on the I2C output anyways.) That could save me around 9mm.

**Reply by dumblob** [reply]
> I could drop the one "conenctor" where I put VCC, GND and a single digital pin. (VCC and GND are replicated on the I2C output anyways.) That could save me around 9mm.

Sounds good to me :+1:. Replicating ports is generally not a good idea IMHO and 9mm should be just about enough to achieve ~56mm width :wink:. Thanks!

**Reply by Dakkaron** [reply]
![grafik](https://user-images.githubusercontent.com/6311558/169768005-00be9fbc-2ad5-4bd4-9013-eea3c20f719c.png)

This is the current revision. The new dimensions are 45.35mm * 16.35mm.

I merged the connectors down into one 2x6 header, which contains a 6-pin-ISP port on the left side, followed by I2C and then 3 Analog pins + AREF. All of the IO pins on that connector can also be used digitally, which gives a total of 8 spare GPIO pins.

Also, I added two LEDs to display status (e.g. Cursor mode or something).

**Reply by dumblob** [reply]
This looks great! I assume you will then [add copper fills/zones](https://store.curiousinventor.com/guides/kicad/pcb_layout/copper/ ) and then perhaps [watch for too tiny copper islands](https://electronics.stackexchange.com/questions/257585/should-we-remove-unconnected-copper-island-among-connected-traces )), right?

The only (and totally unimportant) question is how to distinguish the two LEDs from each other as they are placed very close to each other (a "separating wall" as part of the plastic case would need to be a bit too thin and thus prone to cracking).

**Reply by Dakkaron** [reply]
Thanks for the pointers. As I mentioned, this is my first PCB design project so far, so pointers like that are very welcome.

I don't have a dedicated ground plane, since I wanted to stick with two layers for price reasons.

The LEDs have different colors (blue and yellow). That way, different states should be distingushable. I first considered using an RGB LED, but JLCPCB doesn't have any in their basic parts section.

**Reply by Dakkaron** [reply]
@dumblob 
![grafik](https://user-images.githubusercontent.com/6311558/169838741-92b9cf08-6ff8-4032-817a-47a20cdcde66.png)

Do you like this more?

**Reply by dumblob** [reply]
>I don't have a dedicated ground plane, since I wanted to stick with two layers for price reasons.

Dedicated ground plane is not needed for such low-speed designs. Two layers is more than enough. Just to avoid misunderstanding I really meant copper fills/zones and not a ground plane :wink:.

And yes, LEDs with different colors solve the issue with being close to each other elegantly :wink:. RGB is IMHO unnecessary (JLCPCB though surprised me they don't have it :open_mouth:).

And yes, the latest render with copper fills/zones in the blue layer looks good. Now also copper fills/zones in the red layer and you are probably good to go :wink:.

**Reply by Dakkaron** [reply]
> Dedicated ground plane is not needed for such low-speed designs. Two layers is more than enough. Just to avoid misunderstanding I really meant copper fills/zones and not a ground plane 😉.

Ok, that makes sense, thanks for the input!

> RGB is IMHO unnecessary (JLCPCB though surprised me they don't have it 😮).

Yeah, my thinking was that I'd get more different colours with less space requirement. JLCPCB does have RGB LEDs, but only as Extended Parts. And each different extended part adds an additional $3 setup fee. And that's just not worth the very limited advantage.

> And yes, the latest render with copper fills/zones in the blue layer looks good. Now also copper fills/zones in the red layer and you are probably good to go 😉.

Since there is only very little space on the red layer, and almost all red-fillable areas already overlapped the blue filled areas, I thought it wouldn't do much. I think, I now filled pretty much every single possible space on the board with fill:

![grafik](https://user-images.githubusercontent.com/6311558/169888759-0639715b-9a8a-4254-9d48-a29080df659b.png)

Do you have any other input?

If it's all good, I'll let the design rest for a few days, in case I come up with something else that needs to be changed, and then I'll order 2pcs for a first prototype run.

When that's done and working, I'll see how many people I can get aboard and make a first real batch.

**Reply by dumblob** [reply]
> Do you have any other input?

You asked for it! :laughing: It is actually just nit pics from my side.

1. I thought the clearance from all the edges to the closest trace or SMD piece or a hole or via shall be bigger (but I did not open the design in KiCAD and thus didn't measure how much it is - so it is all guess work from the pictures you posted).

2. All corners (especially the "inner" one near the western corner of the Arduino chip) shall have ~2mm radius (1.5mm radius shall still be acceptable).

3. The traces look a tiny little too thin for cheap manufacturing - but again, I did not measure them, so it is just a huge guess work.

4. The vias look also a tiny little too thin for cheap manufacturing - but again, I did not measure them, so it is just a total guess work on my side.

With (2), (3), and (4) we shall check at JLCPCB (and ideally also one or two others) about their preferences.

> If it's all good, I'll let the design rest for a few days, in case I come up with something else that needs to be changed, and then I'll order 2pcs for a first prototype run.

:+1:

Do I remember correctly that the minimum quantity at JLCPCB was 2 or even more pieces? If so, then feel free to send me a private message on mygithubnick at gmail dot com to find out the cheapest delivery path between our places (I have friends who travel a lot across Europe more often than not).

Do they still also pack some random other PCBs in the package so that one gets a surprise and at the same time can broaden her perspectives among HW?

> When that's done and working, I'll see how many people I can get aboard and make a first real batch.

I'll be one of them (admittedly in my country in Europe 75 EUR is quite a lot, so I'm not sure my friends who wanted to *try* it would consider this *trying* any more and not an *investment* which they would be quite understandably somewhat reluctant to accept).

**Reply by Dakkaron** [reply]
Thanks for the feedback!

(1) I tried to get the clearances as small as KiCAD would let me, to save some space. What kind of clearance should I have?
(2) What do you mean by the radius? Should I draw rounded corners?
(3) I set the traces to the minimum width recommended by JLCPCB
(4) The vias are actually a tiny bit larger than the minimum, that JLCPCB requires

Since the production cost of the PCBs is very small (€1.90 for 5pcs) and the main part of the cost is assembly, it doesn't seem like the traces/vias drive up the price.

> Do I remember correctly that the minimum quantity at JLCPCB was 2 or even more pieces? If so, then feel free to send me a private message on mygithubnick at gmail dot com to find out the cheapest delivery path between our places (I have friends who travel a lot across Europe more often than not).

I'll send you a message.

> Do they still also pack some random other PCBs in the package so that one gets a surprise and at the same time can broaden her perspectives among HW?

Last time I ordered 5 boards and got 25 instead.

> I'll be one of them (admittedly in my country in Europe 75 EUR is quite a lot, so I'm not sure my friends who wanted to try it would consider this trying any more and not an investment which they would be quite understandably somewhat reluctant to accept).

What country do you live in?
If you want to, we can talk per email in a bit more detail in regards of my price considerations.

**Reply by dumblob** [reply]
> (1) I tried to get the clearances as small as KiCAD would let me, to save some space. What kind of clearance should I have?

IDK how much it is in your design but to be on the "safe side" I would probably choose 1.5 mm (it highly depends on the manufacturing process where they cut it along straight edges etc. and on the mounting in the smartphone case where you might want to have some "mounting click supports" around the PCB as shown below).

![C6-PCB-Mount-Box](https://user-images.githubusercontent.com/2451044/169969406-1154d41f-5631-4182-80ca-9501f310eaf7.jpg)

>(2) What do you mean by the radius? Should I draw rounded corners?

Yes. JLCPCB will anyway make a radius on the one inner corner near the arduino chip. So it is better to define the radius by yourself than to let them choose one for you :wink:.

>(3) I set the traces to the minimum width recommended by JLCPCB

Hm, I for myself wouldn't go for the minimum width (I know etching nowadays is very good but something from the past still tells me it might be unnecessarily risky) but YMMW :wink:.

>(4) The vias are actually a tiny bit larger than the minimum, that JLCPCB requires

Ok. Usually PCB shops tell which vias are easy to drill (they drill several PCBs at once with thicker drills unlike with thin drills and thus the price goes quickly up). I did not check JLCPCB.

> Since the production cost of the PCBs is very small (€1.90 for 5pcs) and the main part of the cost is assembly, it doesn't seem like the traces/vias drive up the price.

The width of traces should not influence price - it's really about reliability of the final PCB. Vias diameter though usually does influence the price - but maybe not at JLCPCB. So if it doesn't in this case, then leave them as they are :wink:.

> I'll send you a message.

I got it and will reply soon. Thanks!

> Last time I ordered 5 boards and got 25 instead.

:rofl:

**Comment by spARTacusFromCB** [comment]
So the latest thought is the board will be no wider than the actual Q10 Keyboard? Do the latest thoughts include extra points on the board for if someone in the future wanted to explore adding some extra keys/trackpad/belt? I realize there probably isn't suitable space above the keyboard for a belt, but my Samsung A21 is a bit wider than a Q10 Keyboard. So, some folks might be able to squeeze in extras on the left / right sides of the keyboard, and still not exceed the width of the phone.

**Reply by dumblob** [reply]
Yes, that is exactly the reason why we try to make the new PCB as small and narrow as possible. You can then easily add other PCBs (with buttons, joystics, etc.) and solder them (either using a wire or using castellated holes) to the new PCB from here (I am privately in contact with @Dakkaron where we discuss using pads instead of through-hole connections - if pads will be used in the end - and I hope they will - then castellated soldering is possible unlike with the new PCB revision above).

Or you can take the new PCB layout and simply extend it in KiCAD in any direction you want and just place the buttons etc. there (because it is cheap to order assembled e.g. from JLCPCB).

So again, making the PCB smaller covers all the possible future use cases.
